File: frontend/src/services/recurring.ts
// frontend/src/services/recurring.ts
import API from './api'
import type { RecurringRead, RecurringCreate, RecurringUpdate } from '../types'

/* ── Fetch all recurring transactions ─────────────────── */
export function fetchRecurring(): Promise<RecurringRead[]> {
  return API.get<RecurringRead[]>('/recurring').then(r => r.data)
}

/* ── Add recurring transaction ─────────────────── */
export function createRecurring(r: RecurringCreate): Promise<RecurringRead> {
  return API.post<RecurringRead>('/recurring', r).then(rsp => rsp.data)
}

/* ── Update recurring ────────────────────── */
export function updateRecurring(
  id: number,
  r: RecurringUpdate
): Promise<RecurringRead> {
  return API.patch<RecurringRead>(`/recurring/${id}`, r).then(rsp => rsp.data)
}

/* ── Remove recurring transaction ──────────────────── */
export function deleteRecurring(id: number): Promise<void> {
  return API.delete<void>(`/recurring/${id}`).then(() => {})
}

File: backend/app/routers/recurring.py
#backend/app/routers/recurring.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel.ext.asyncio.session import AsyncSession
from sqlmodel import select

from app.models import RecurringTransaction
from app.schemas.recurring import RecurringCreate, RecurringRead, RecurringUpdate
from app.core.security import get_current_user
from app.db import get_session

router = APIRouter(tags=["recurring"])

@router.post("", response_model=RecurringRead, status_code=status.HTTP_201_CREATED)
async def create_recurring(
    rec_in: RecurringCreate,
    current=Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    r = RecurringTransaction(**rec_in.dict(), user_id=current.user_id)
    session.add(r)
    await session.commit()
    await session.refresh(r)
    return r

@router.get("", response_model=List[RecurringRead])
async def read_recurrings(
    current=Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    result = await session.exec(
        select(RecurringTransaction).where(RecurringTransaction.user_id == current.user_id)
    )
    return result.all()

@router.patch("/{rec_id}", response_model=RecurringRead)
async def update_recurring(
    rec_id: int,
    updates: RecurringUpdate,
    current=Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    r = await session.get(RecurringTransaction, rec_id)
    if not r or r.user_id != current.user_id:
        raise HTTPException(status_code=404, detail="Recurring not found")
    for k, v in updates.dict(exclude_unset=True).items():
        setattr(r, k, v)
    await session.commit()
    await session.refresh(r)
    return r

@router.delete("/{rec_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_recurring(
    rec_id: int,
    current=Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
):
    r = await session.get(RecurringTransaction, rec_id)
    if not r or r.user_id != current.user_id:
        raise HTTPException(status_code=404, detail="Recurring not found")
    await session.delete(r)
    await session.commit()
